Incorporate DSL2 features into docs (#3793)
nextflow-io/nextflow: docs/module.rst

.. _module-page:

*******
Modules
*******

.. note::
    Modules were introduced in DSL2. If you are still using DSL1, see the :ref:`dsl1-page` page to
    learn how to migrate your Nextflow pipelines to DSL2.

> 

Reviewer: I wonder if this is even needed. Maybe keep it, but drop it down below the _Functions, processes, and workflows_ note below? Just so that it's not the first thing you see on the page.

Author: Fair enough

.. _module-directory:

Module directory
----------------

.. note::
    This feature requires Nextflow version ``22.10.0`` or later.

A module can be defined as a directory with the same name as the module and with a script
named ``main.nf``. For example::

    some
        └-module
            └-main.nf

When defined as a directory, the module must be included by specifying the module directory path::

    include { foo } from './some/module'

Module directories allow the use of module scoped binaries scripts. See `Module binaries`_ for details.

Multiple inclusions
-------------------

A Nextflow script can include any number of modules, and an ``include`` statement can import any number of
components from a module. Multiple components can be included from the same module by using the syntax
shown below::

    include { foo; bar } from './some/module'

    workflow {
        data = channel.fromPath('/some/data/*.txt')
        foo(data)
        bar(data)
    }

.. _module-aliases:

Module aliases
--------------

When including a module component, it's possible to specify an *alias* with the ``as`` keyword.

> Aliasing allows you to include and invoke components with the same name, by assigning them different

Reviewer: ```suggestion
Aliasing allows you to avoid module name clashes, by assigning them different
```

names in the including context. For example::

    include { foo } from './some/module'
    include { foo as bar } from './other/module'

    workflow {
        foo(some_data)
        bar(other_data)
    }

You can even include the same component multiple times under different names::

>     include { foo; foo as bar } from './some/module'

Reviewer: Ooh nice, I've never seen this done in one line before! I have just duplicated the include line a bunch of times.. 🤩

Author: *furiously updates all of the include statements across nf-core*

    workflow {
        foo(some_data)
        bar(other_data)
    }

Module parameters
-----------------

A module script can define parameters using the same syntax as a Nextflow workflow script::

    params.foo = 'Hello'
    params.bar = 'world!'

    def sayHello() {
        println "$params.foo $params.bar"
    }

When including a module, the module will first use parameters from the including context. For example::

    params.foo = 'Hola'
    params.bar = 'Mundo'

    include { sayHello } from './some/module'

    workflow {
        sayHello()
    }

The above snippet prints::

    Hola Mundo

.. note::
    The module inherits the parameters defined *before* the ``include`` statement, therefore any parameters
    set afterwards will not be used by the module.

.. tip::
    It is best to define all pipeline parameters *before* any ``include`` statements.

The ``addParams`` option can be used to pass parameters to the module without affecting the including

> scope.

Reviewer: I find this wording a little awkward, but can't think of a better way to say it 🤔

Author: I will give it a try :smile:

                |─P2-template.sh
        └─Phase1-Modules
            |─mymodules3.nf
            |─mymodules4.nf
            └─templates
                |─P3-template.sh
                └─P4-template.sh
        └─Phase2-Modules
            |─mymodules5.nf
            |─mymodules6.nf
            └─templates
                |─P5-template.sh
                |─P6-template.sh
                └─P7-template.sh

Module binaries
---------------

.. note::
    This feature requires Nextflow version ``22.10.0`` or later.

Modules can define binary scripts that are locally scoped to the processes defined by the tasks.

To enable this feature, enable the following flag in your pipeline script or configuration file::

    nextflow.enable.moduleBinaries = true

The binary scripts must be placed in the module directory names ``<module-dir>/resources/usr/bin``::

    <module-dir>
        |─main.nf
        └─resources
            └─usr
                └─bin
                    |─your-module-script1.sh
                    └─another-module-script2.py

Those scripts will be made accessible like any other command in the task environment, provided they have been granted
the Linux execute permissions.

> .. note::
>     This feature requires the use of a local or shared file system for the pipeline work directory, or
>     :ref:`wave-page` when using container-native executors.

Reviewer: Would be good to explicitly say what _won't_ work here. Not entirely sure if we're saying that it doesn't work with Docker? Or bucket storage? Or both?

Also should be a warning and not a note.

```suggestion
.. warning::
    This feature requires the use of a local or shared file system for the pipeline work directory.
    It will not work with container technologies such as Docker and work directories stored in buckets.
    To use module binaries in these situations you can use :ref:`wave-page`.
```

Author: This one is my bad, originally it said "cloud-based executors" and I changed it to "container-native" because I thought it was more accurate, but now I think I was wrong. I will change it back to "cloud-based".
